test/wasm_test.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "include/proxy-wasm/wasm.h"
+
+#include "gtest/gtest.h"
+
+#include "test/utility.h"
+
+namespace proxy_wasm {
+
+auto test_values = testing::ValuesIn(getRuntimes());
+
+INSTANTIATE_TEST_SUITE_P(Runtimes, TestVM, test_values);
+
+// Failcallbacks only used for runtimes - not available for nullvm.
+TEST_P(TestVM, GetOrCreateThreadLocalWasmFailCallbacks) {
+  const auto plugin_name = "plugin_name";
+  const auto root_id = "root_id";
+  const auto vm_id = "vm_id";
+  const auto plugin_config = "plugin_config";
+  const auto fail_open = false;
+
+  // Create a plugin.
+  const auto plugin = std::make_shared<PluginBase>(plugin_name, root_id, vm_id, runtime_,
+                                                   plugin_config, fail_open, "plugin_key");
+
+  // Define callbacks.
+  WasmHandleFactory wasm_handle_factory =
+      [this](std::string_view vm_key) -> std::shared_ptr<WasmHandleBase> {
+    auto base_wasm = std::make_shared<WasmBase>(newVm(), "vm_id", "vm_config", vm_key,
+                                                std::unordered_map<std::string, std::string>{},
+                                                AllowedCapabilitiesMap{});
+    return std::make_shared<WasmHandleBase>(base_wasm);
+  };
+
+  WasmHandleCloneFactory wasm_handle_clone_factory =
+      [this](std::shared_ptr<WasmHandleBase> base_wasm_handle) -> std::shared_ptr<WasmHandleBase> {
+    auto wasm = std::make_shared<WasmBase>(base_wasm_handle,
+                                           [this]() -> std::unique_ptr<WasmVm> { return newVm(); });
+    return std::make_shared<WasmHandleBase>(wasm);
+  };
+
+  PluginHandleFactory plugin_handle_factory =
+      [](std::shared_ptr<WasmHandleBase> base_wasm,
+         std::shared_ptr<PluginBase> plugin) -> std::shared_ptr<PluginHandleBase> {
+    return std::make_shared<PluginHandleBase>(base_wasm, plugin);
+  };
+
+  // Read the minimal loadable binary.
+  auto source = readTestWasmFile("abi_export.wasm");
+
+  // Create base Wasm via createWasm.
+  auto base_wasm_handle =
+      createWasm("vm_key", source, plugin, wasm_handle_factory, wasm_handle_clone_factory, false);
+  ASSERT_TRUE(base_wasm_handle && base_wasm_handle->wasm());
```

Comment:
I think the `base_wasm_handle` will always point to the original failed VM, and never to a restarted VM, which is kind of a dangling pointer/handle issue.

I'm not sure if this can lead to any real issues in the existing codebase, but it's something to think about.

Reply:
well, the failed VMs are cloned ones from this `base_wasm_handle`, so I don't think this always points to the failed one. Added assertions for this: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/181/commits/f498a339947fbcc55f05fc06c044bb568879713e

Reply:
Sorry, my mistake. Not sure why I thought it points to the thread-local version.